src/theory/shared_solver.cpp

```diff
@@ -0,0 +1,116 @@
+/*********************                                                        */
+/*! \file shared_solver.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief The shared solver base class
+ **/
+
+#include "theory/shared_solver.h"
+
+#include "expr/node_visitor.h"
+#include "theory/theory_engine.h"
+
+namespace CVC4 {
+namespace theory {
+
+// Always creates shared terms database. In all cases, shared terms
+// database is used as a way of tracking which calls to Theory::addSharedTerm
+// we need to make in preNotifySharedFact.
+// In distributed equality engine management, shared terms database also
+// maintains an equality engine. In central equality engine management,
+// it does not.
+SharedSolver::SharedSolver(TheoryEngine& te)
+    : d_te(te),
+      d_logicInfo(te.getLogicInfo()),
+      d_sharedTerms(&d_te, d_te.getSatContext()),
+      d_sharedTermsVisitor(d_sharedTerms)
+{
+}
+
+bool SharedSolver::needsEqualityEngine(theory::EeSetupInfo& esi)
+{
+  return false;
+}
+
+void SharedSolver::preRegisterShared(TNode t, bool multipleTheories)
+{
+  // register it with the equality engine manager if shared is enabled
+  if (d_logicInfo.isSharingEnabled())
+  {
+    preRegisterSharedInternal(t);
+  }
+  // if multiple theories are present in t
+  if (multipleTheories)
+  {
+    // Collect the shared terms if there are multiple theories
+    // This calls Theory::addSharedTerm, possibly multiple times
+    NodeVisitor<SharedTermsVisitor>::run(d_sharedTermsVisitor, t);
```

Comment:
Eventually, we should consider moving the `SharedTermsVisitor` to this source file since it is (presumably) not going to be used anywhere else.

Reply:
Right, this also the case for parts of SharedTermsDatabase.